### ISSUE TITLE
[prefer-readonly-parameter-types] False positives with nested object

### ISSUE THREAD

**Issue by C0ZEN** [typescript-eslint/typescript-eslint#2823]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/prefer-readonly-parameter-types": [
		"error", {
			"ignoreInferredTypes": true, 
			"checkParameterProperties": false
		}
	]
  }
}
```

```TS
interface IEnvironment {
  sub: IEnvironmentSub;
}

interface IEnvironmentSub {
  yo: string;
}


class Dummy {
  public logEnv(env: Readonly<IEnvironment>): void {
    console.log(env);
  }
}
```

**Expected Result**

No linter error.

**Actual Result**

```
ESLint: Parameter should be a read only type.(@typescript-eslint/prefer-readonly-parameter-types)
```

**Additional Info**

Changing `sub` to `sub?` fix the error...

```TS
interface IEnvironment {
  sub?: IEnvironmentSub;
}
```

**Versions**

| package                            | version  |
| ---------------------------------- | -------- |
| `@typescript-eslint/eslint-plugin` | `4.8.2`  |
| `@typescript-eslint/parser`        | `4.8.2`  |
| `TypeScript`                       | `4.1.2`  |
| `ESLint`                           | `7.14.0` |
| `node`                             | `14.3.0` |

**Comment by bradzacher** [comment]
https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/prefer-readonly-parameter-types.md#rule-details

> A type is considered readonly if:
> ...
> it is an object type whose properties are all marked as readonly, and whose values are all considered readonly.

```ts
interface IEnvironment {
  sub: IEnvironmentSub;
}

interface IEnvironmentSub {
  yo: string;
}


class Dummy {
  public logEnv(env: Readonly<IEnvironment>): void {
    console.log(env);
    env.sub = { yo: 'a' }; // Expected Error - Cannot assign to 'sub' because it is a read-only property

    env.sub.yo = 'b'; // Expected error but got none
  }
}
```
[repl](https://www.typescriptlang.org/play?#code/JYOwLgpgTgZghgYwgAgJIFEQDdhQPYgC2E4yA3gLABQyyAzgK4BGAXGpjvkSWAMrMBuagF9q1UJFiIUGbLgLFw-JuWq0AnnjZ0wUUAHMhVUVTFUEAGzh06yACINChdaprIADswvAEyC3n0OAAoSLDYAJQg4ABMCC3UAHllOBR4APgBKNiw8YGjXWloEAjo8CwgAOn99EOwMo0LkUIrGFQBecmRNNgByOB7kYQFkAHoR5HQAD3cIBEh89Ch8KGQAWmQAYTgQEDwwZGs6YH0QZDA8ZB7WgaZZuAY6FGB94Fs4ZCgo6NW4l3d8GZQMDqMyNZqtCqaZAdHpMHrDMYTaazeZNJZ4FZMBj7fR7ZC7EAQNSDERAA)

Your type is not deeply readonly - only the first level is readonly.
You need to provide a deeply readonly type for the rule to be satisfied.
